cloudinit/templater.py

```diff
@@ -102,18 +111,72 @@ def detect_template(text):
     def jinja_render(content, params):
         # keep_trailing_newline is in jinja2 2.7+, not 2.6
         add = "\n" if content.endswith("\n") else ""
-        return (
-            JTemplate(
-                content,
-                undefined=UndefinedJinjaVariable,
-                trim_blocks=True,
-                extensions=["jinja2.ext.do"],
-            ).render(**params)
-            + add
-        )
+        try:
+            result = (
+                JTemplate(
+                    content,
+                    undefined=UndefinedJinjaVariable,
+                    trim_blocks=True,
+                    extensions=["jinja2.ext.do"],
+                ).render(**params)
+                + add
+            )
+            return result
+
+        except Exception as jtemplate_error:
+            # if the error is a syntax one, we know this next line will fail
+            # we want to get the traceback and error message it throws
+            try:
+                jinja_env = Environment(
+                    loader=BaseLoader(),
+                    autoescape=False,
+                    undefined=UndefinedJinjaVariable,
+                    trim_blocks=True,
+                    extensions=["jinja2.ext.do"],
+                )
+                jinja_env.parse(content)
+                # if we get here, then the previous line didn't fail,
+                # so we just need to re-raise the original error
+                raise jtemplate_error
+            except TemplateSyntaxError as template_syntax_error:
+                try:
```

Comment:
As we are now capturing `TemplateSyntaxError`, we could potentially simplify this logic involving regexes to just using [TemplateSyntaxError.lineno](https://jinja.palletsprojects.com/en/3.1.x/api/#jinja2.TemplateSyntaxError.lineno).